components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java

```diff
@@ -1309,6 +1330,223 @@ public void addApplicationWithTemplateIdAndTemplateVersionData(String templateId
+        CertificateMgtClientException clientException = new CertificateMgtClientException(invalidCertExp.getMessage(),
+                invalidCertExp.getDescription(), invalidCertExp.getCode());
+
+        // Create an application with certificate.
+        doReturn(CERTIFICATE_ID).when(applicationCertificateManagementService).addCertificate(any(), anyString());
+        String resourceId = applicationManagementService.createApplication(inputSP, SUPER_TENANT_DOMAIN_NAME,
+                REGISTRY_SYSTEM_USERNAME);
+        //  Retrieving added application.
+        doReturn(certificate).when(applicationCertificateManagementService).getCertificate(anyInt(), anyString());
+        ServiceProvider retrievedSP = applicationManagementService.getApplicationByResourceId(resourceId,
+                SUPER_TENANT_DOMAIN_NAME);
+
+        // Update application with client error when updating certificate.
+        doThrow(clientException).when(applicationCertificateManagementService).updateCertificateContent(anyInt(),
+                any(), anyString());
+        try {
+            retrievedSP.setCertificateContent(UPDATED_CERTIFICATE);
+            applicationManagementService.updateApplicationByResourceId(resourceId, retrievedSP,
+                    SUPER_TENANT_DOMAIN_NAME, REGISTRY_SYSTEM_USERNAME);
+            Assert.fail();
+        } catch (IdentityApplicationManagementException e) {
+            Assert.assertEquals(e.getClass(), IdentityApplicationManagementClientException.class);
+            Assert.assertEquals(e.getMessage(), invalidCertExp.getDescription());
+        }
+
+        // Deleting added application.
+        doNothing().when(applicationCertificateManagementService).deleteCertificate(anyInt(), anyString());
+        applicationManagementService.deleteApplication(inputSP.getApplicationName(),
+                SUPER_TENANT_DOMAIN_NAME, REGISTRY_SYSTEM_USERNAME);
+
+        // Add application with client error when adding certificate.
+        doThrow(clientException).when(applicationCertificateManagementService).addCertificate(any(), anyString());
+        try {
+            inputSP = new ServiceProvider();
+            inputSP.setApplicationName(APPLICATION_NAME_1);
+            inputSP.setCertificateContent(CERTIFICATE);
+            applicationManagementService.createApplication(inputSP, SUPER_TENANT_DOMAIN_NAME, REGISTRY_SYSTEM_USERNAME);
+            Assert.fail();
+        } catch (IdentityApplicationManagementException e) {
+            Assert.assertEquals(e.getClass(), IdentityApplicationManagementClientException.class);
+            Assert.assertEquals(e.getMessage(), invalidCertExp.getDescription());
+        }
+    }
+
+    @Test(dataProvider = "applicationCertificateDataProvider", dependsOnMethods = "testApplicationCertificateSuccess")
+    public void testApplicationCertificateServerError(ServiceProvider inputSP, Certificate certificate)
```

Comment:
Can we break these add, update retrieve test cases to different methods

Reply:
Fixed with https://github.com/wso2/carbon-identity-framework/pull/6087/commits/9a2dc889d25ec1c2d6d950643a6310f1717fd44a